chain/client/src/client.rs

```diff
@@ -807,6 +810,166 @@ impl Client {
         }
     }
 
+    /// Processes received block. Ban peer if the block header is invalid or the block is ill-formed.
+    // This function is just a wrapper for process_block_impl that makes error propagation easier.
+    pub fn receive_block(
+        &mut self,
+        block: Block,
+        peer_id: PeerId,
+        was_requested: bool,
+        apply_chunks_done_callback: DoneApplyChunkCallback,
+    ) {
+        let hash = *block.hash();
+        let prev_hash = *block.header().prev_hash();
+        let _span = tracing::debug_span!(
+            target: "client",
+            "process_block",
+            me = ?self.validator_signer.as_ref().map(|vs| vs.validator_id()),
+            prev_hash = %prev_hash,
+            %hash,
+            height = block.header().height(),
+            %peer_id,
+            was_requested)
+        .entered();
+
+        match self.receive_block_impl(
+            block,
+            peer_id.clone(),
+            was_requested,
+            apply_chunks_done_callback,
+        ) {
+            Ok(_) => {}
+            // Log the errors here. Note that the real error handling logic is already
+            // done within process_block_impl, this is just for logging.
+            Err(ref err) if err.is_bad_data() => {
+                warn!(target: "client", "Receive bad block: {}", err);
+            }
+            Err(ref err) if err.is_error() => {
+                if let near_chain::Error::DBNotFoundErr(msg) = err {
+                    debug_assert!(!msg.starts_with("BLOCK HEIGHT"), "{:?}", err);
+                }
+                if self.sync_status.is_syncing() {
+                    // While syncing, we may receive blocks that are older or from next epochs.
+                    // This leads to Old Block or EpochOutOfBounds errors.
+                    debug!(target: "client", "Error on receival of block: {}", err);
+                } else {
+                    error!(target: "client", "Error on receival of block: {}", err);
+                }
+            }
+            Err(e) => {
+                debug!(target: "client", error = %e, "Process block: refused by chain");
+            }
+        }
+    }
+
+    /// Processes received block.
+    /// This function first does some pre-check based on block height to avoid processing
+    /// blocks multiple times.
+    /// Then it process the block header. If the header if valid, broadcast the block to its peers
+    /// Then it starts the block processing process to process the full block.
+    pub(crate) fn receive_block_impl(
+        &mut self,
+        block: Block,
+        peer_id: PeerId,
+        was_requested: bool,
+        apply_chunks_done_callback: DoneApplyChunkCallback,
+    ) -> Result<(), near_chain::Error> {
+        let prev_hash = *block.header().prev_hash();
```

Comment:
move prev-hash calculation closer to the use?

Reply:
`block` is moved later so prev_hash has to be calculated before that